### DISCUSSION TITLE
Invite users via e-mail for 3 level SaaS scenario

### DISCUSSION THREAD

**Discussion by narcis-ro** [zitadel/zitadel#5563] [General]
Hello. We’re doing a PoC with Zitadel Cloud to see if we can integrare it into our SaaS product. Penny for your thoughts, please bear with us through this long post :)

We have the following hierarchy: Tenants > Business Accounts > Distributors

One user can have access to mutiple tenants, business accounts, distributors and can have specific roles to each area. 

Tenant:
   owner role:
          -  can create business accounts, assigns a business account an owner
          -  can invite other users, with assigned roles (specific at tenant level)
          
Business Account:
    owner role:
           - can create distributors, assigns a distributor owner
           - can invite other users, with assigned roles in his business account, specific business account roles

Distributors:
    owner role:
          - can invite other users with assigned role specific at distributor level, roles specific to distributors

Each user can invite another user via e-mail and grant a specific role in that organization unit (tenant, business account, distributor). The user would receive thie e-mail which probably contains a invitation code. He would be redirected to one of our app, in which the user is requested to login or register if it doesn’t have an identity in our platform yet.

After login or register we would rediect the user to claim the invitation (hopefully making use of the oidc state to keep the invitation code cross redirects). 
At this point we would probably make a api call to zitadel to add roles to the user at the invited organization unit (tenant, business account, distributor), but since Zitadel only has 2 levels, organizations and projects. We were wondering how to best model this?

If we were to keep roles in the access token. After the invite is claimed we would need to refresh the claims in the token. Can that be done one the fly or the user needs to logout/login?
Another question is if we should give access to tenants in the Zitadel console to manage their own users and assigned roles, same for business accounts and distributors. Although a lot of options should not be available to them, so they don’t shoot themselves in the foot. 

I guess our SaaS kinda looks like your Zitadel Cloud model. You have clients (tenants in our case), and your clients have 2 level of organizing via organizations and projects (business accounts and distributorson our side).

Other info:
- although a user can have access to multiple tenants, business accounts and distributors, they won’t have more than 5-10
- There are 3 applications; One is admin spa portal for managing tenant settings and creating business accounts, another one is partner spa portal, in which a business accounts can login and create distributors and peform other actions. And a mobile app, same as an extension of the partner portal but dedicated to field operations.
- Logout/Login for roles change is acceptable but undesired if possible
- We plan to use AzureAD for our own employees and user/pass+2fa for other users.

**Comment by fforootd** [comment]
Good questions let me try and respond as good as I understand this.

> One user can have access to mutiple tenants, business accounts, distributors and can have specific roles to each area.

Generally we would recommend to create an organization for each "Business Partner" and then to grant them the roles of your projects.
We call this an [project grant](https://zitadel.com/docs/concepts/usecases/saas#project-grant).

> After login or register we would rediect the user to claim the invitation (hopefully making use of the oidc state to keep the invitation code cross redirects).

Sound like a good plan. You can define a default redirect in zitadel as well.

> At this point we would probably make a api call to zitadel to add roles to the user at the invited organization unit (tenant, business account, distributor), but since Zitadel only has 2 levels, organizations and projects. We were wondering how to best model this?

You could also create an action to fetch these infos of an http api. E.g during register/login you could fetch a json from your api that represents roles and or additional metadata. (this works also on zitadel's data)

> If we were to keep roles in the access token. After the invite is claimed we would need to refresh the claims in the token. Can that be done one the fly or the user needs to logout/login?

Usually that needs a login, but that can be done silently or without user intervention with the use of the [prompt parameter](https://zitadel.com/docs/apis/openidoauth/endpoints#additional-parameters)

Our recommend way would be to use the /userinfo and /introspect endpoint to fetch the roles at runtime from ZITADEL.
You then can cache the response for a set amount of time which gives you the same benefits as with JWT tokens but with more flexibility.

> I guess our SaaS kinda looks like your Zitadel Cloud model. You have clients (tenants in our case), and your clients have 2 level of organizing via organizations and projects (business accounts and distributorson our side).

I guess your case looks kind of the same. We have tenants (in our zitadel each customer is an org.) and we give the roles by project grants depeding on their needs. (So that we could delegate more access to a partner then a customer for example)

Hope this helps, otherwise let me know
